### PR TITLE
Stream wrapper + tests

### PR DESCRIPTION
**Pull request by bretambrose**
Converts from basic C++ streams into aws-c streams, allowing users to use C++ streams as http request bodies.

This lets us give a C++ stream interface to the Http Request wrapper.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/crt/Types.h [hunk 1]
```diff
@@ -45,7 +45,8 @@ namespace Aws
         namespace Io
         {
             using SocketOptions = aws_socket_options;
-        }
+            using IoStream = std::basic_iostream<char, std::char_traits<char>>;
```
Comment (JonathanHenson) [inline include/aws/crt/Types.h]: is an allocator needed here?

File: include/aws/crt/io/Stream.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            /*
+             * Factory to create a aws-c-io input stream subclass from a C++ stream
+             */
+            AWS_CRT_CPP_API aws_input_stream *AwsInputStreamNewCpp(
+                const std::shared_ptr<Aws::Crt::Io::IoStream> &stream,
+                Aws::Crt::Allocator *allocator = DefaultAllocator()) noexcept;
+        } // namespace Io
+    }     // namespace Crt
+} // namespace Aws
```
Comment (JonathanHenson) [inline include/aws/crt/io/Stream.h]: newline

File: source/io/Stream.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                input_stream->vtable = &s_aws_input_stream_cpp_vtable;
+                input_stream->impl = impl;
+
+                impl->stream = stream;
+
+                return input_stream;
+            }
+        } // namespace Io
+    }     // namespace Crt
+} // namespace Aws
```
Comment (JonathanHenson) [inline source/io/Stream.cpp]: newline

File: tests/StreamTest.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            STREAM_CONTENTS + strlen(STREAM_CONTENTS) + END_SEEK_OFFSET, read, buffer.buffer, read);
+
+        aws_byte_buf_clean_up(&buffer);
+        aws_input_stream_destroy(wrapped_stream);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(StreamTestSeekEnd, s_StreamTestSeekEnd)
```
Comment (JonathanHenson) [inline tests/StreamTest.cpp]: newline

File: source/io/Stream.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                struct aws_input_stream *input_stream = NULL;
+                struct aws_input_stream_cpp_impl *impl = NULL;
+
+                aws_mem_acquire_many(
+                    allocator,
+                    2,
+                    &input_stream,
+                    sizeof(struct aws_input_stream),
+                    &impl,
+                    sizeof(struct aws_input_stream_cpp_impl));
```
Comment (ColdenCullen) [inline source/io/Stream.cpp]: Yay, a good use of acquire_many! Love it.

File: include/aws/crt/Types.h (hunk 1 above)
Comment (ColdenCullen) [reply include/aws/crt/Types.h]: [no](https://en.cppreference.com/w/cpp/io/basic_iostream)

Comment (ColdenCullen) [inline include/aws/crt/Types.h]: Any reason to not just use `std::basic_istream`? We're never writing to this thing.

Comment (bretambrose) [reply include/aws/crt/Types.h]: Probably not.  We'll eventually need an output stream as well, but this can be changed.

**Pull request merged**